### ISSUE TITLE
Function parameter lost

### ISSUE THREAD

**Issue by ThaDafinser** [gkralik/php7-sapnwrfc#15]
It seems that the function invoke parameters are lost

# PHP7
```php
$conn = new SapConnection($config);

$parameters = [
    'FMI_MATNR' => '887070'
];

$func = $conn->getFunction('Z_PPP_MAT_GET_METADATA');
$result = $func->invoke($parameters);
var_dump($result);
```

## Trace
```
>> RfcOpenConnection
>> RfcOptions
	// ...
<< RfcOptions

>> RfcConnection name="none"
Handle = 108726000 ConvID = 42090651
<< RfcConnection

<< RfcOpenConnection returned RFC_OK

>> RfcGetFunctionDesc
RfcGetFunctionDesc(Z_PPP_MAT_GET_METADATA) via handle 108726000
RfcGetFunctionDesc(Z_PPP_MAT_GET_METADATA) via handle 108726000 (SID=T11) returned 00000000067237F0
<< RfcGetFunctionDesc returned RFC_OK

>> RfcInvoke
Invoking Z_PPP_MAT_GET_METADATA on handle 108726000

>> RfcFunction name="Z_PPP_MAT_GET_METADATA"
>> IMPORTING parameters
>> FMI_LAEDA, RFCTYPE_DATE, Length: 16, Active: false, Requested: false
	Value: 0000-00-00
<< FMI_LAEDA

>> FMI_MATNR, RFCTYPE_CHAR, Length: 36, Active: false, Requested: false
	Value: 
<< FMI_MATNR

>> FMI_MTART, RFCTYPE_CHAR, Length: 8, Active: false, Requested: false
	Value: 
<< FMI_MTART

<< IMPORTING parameters

>> CHANGING parameters
<< CHANGING parameters

```

# PHP5
```php
$conn = new sapnwrfc($config);

$parameters = [
    'FMI_MATNR' => '887070'
];

$func = $conn->function_lookup('Z_PPP_MAT_GET_METADATA');
$result = $func->invoke($parameters);
var_dump($result);
```

## Trace
```
>> RfcOpenConnection
>> RfcOptions
	// ...
<< RfcOptions

>> RfcConnection name="none"
Handle = 22353560 ConvID = 42046954
<< RfcConnection

<< RfcOpenConnection returned RFC_OK

>> RfcGetFunctionDesc
RfcGetFunctionDesc(Z_PPP_MAT_GET_METADATA) via handle 22353560
RfcGetFunctionDesc(Z_PPP_MAT_GET_METADATA) via handle 22353560 (SID=T11) returned 015487C8
<< RfcGetFunctionDesc returned RFC_OK

>> RfcInvoke
Invoking Z_PPP_MAT_GET_METADATA on handle 22353560

>> RfcFunction name="Z_PPP_MAT_GET_METADATA"
>> IMPORTING parameters
>> FMI_LAEDA, RFCTYPE_DATE, Length: 16, Active: true, Requested: false
	Value: 0000-00-00
<< FMI_LAEDA

>> FMI_MATNR, RFCTYPE_CHAR, Length: 36, Active: true, Requested: false
	Value: 887070
<< FMI_MATNR

>> FMI_MTART, RFCTYPE_CHAR, Length: 8, Active: true, Requested: false
	Value: 
<< FMI_MTART

<< IMPORTING parameters

>> CHANGING parameters
<< CHANGING parameters
```

**Comment by gkralik** [comment]
Seems they are not active (`[...] Active: false [...]`) strange.

**Comment by gkralik** [comment]
Try calling `$func->setParameterActive('FMI_MATNR', true)` and check if it works.
There is a known bug with parameter activatation, I'll have a look...

**Comment by ThaDafinser** [comment]
Works.
```
>> RfcOpenConnection
>> RfcOptions
	// ...
<< RfcOptions

>> RfcConnection name="none"
Handle = 108726000 ConvID = 43018472
<< RfcConnection

<< RfcOpenConnection returned RFC_OK

>> RfcGetFunctionDesc
RfcGetFunctionDesc(Z_PPP_MAT_GET_METADATA) via handle 108726000
RfcGetFunctionDesc(Z_PPP_MAT_GET_METADATA) via handle 108726000 (SID=T11) returned 00000000067237F0
<< RfcGetFunctionDesc returned RFC_OK

>> RfcInvoke
Invoking Z_PPP_MAT_GET_METADATA on handle 108726000

>> RfcFunction name="Z_PPP_MAT_GET_METADATA"
>> IMPORTING parameters
>> FMI_LAEDA, RFCTYPE_DATE, Length: 16, Active: false, Requested: false
	Value: 0000-00-00
<< FMI_LAEDA

>> FMI_MATNR, RFCTYPE_CHAR, Length: 36, Active: true, Requested: false
	Value: 887070
<< FMI_MATNR

>> FMI_MTART, RFCTYPE_CHAR, Length: 8, Active: false, Requested: false
	Value: 
<< FMI_MTART

<< IMPORTING parameters

>> CHANGING parameters
<< CHANGING parameters
```

**Comment by gkralik** [comment]
OK, use that as a workaround for now. I hope to find enough time to rework the parameter handling this week. I'll keep you posted.

Thanks for testing :+1:

**Comment by ThaDafinser** [comment]
Maybe just call `$func->setParameterActive('FMI_MATNR', true)` when a parameter is provided?

NO problem. Testing is important to get it into production asap ;-)

**Comment by gkralik** [comment]
Yes, the plan is to have a `HashTable` save the activation state and activate each parameter before invoking the RFC.

**Comment by gkralik** [comment]
Can't really reproduce this one either... all parameters seem to be active by default.

**Comment by ThaDafinser** [comment]
Hmm...cant provide you more then the trace and the used code currently...

**Comment by gkralik** [comment]
No problem.

**Comment by gkralik** [comment]
I can't really reproduce your issue. For me, all parameters are active by default. Calling `setParameterActive()` for all FM parameters seems not to be a solution as it causes segfaults when exceptions occur in the FM...
But there have been changes to parameter handling, so maybe your issue is already resolved.

**Comment by gkralik** [comment]
OK, the new implementation now makes sure that all parameters passed as arguments to `invoke()` are activated. This seems to work and does not cause any strange errors.

**Comment by gkralik** [comment]
fixed in 0.7.0
